### PR TITLE
propagate the dockerd cgroup-parent config to buildkitd

### PR DESCRIPTION
**Pull request by AntaresS**
**- What I did**
- propagate the dockerd cgroup-parent from daemon config to buildkitd
  - refactored `new buildkit` creation into `daemon_unix.go` and `daemon_windows.go` for cross platform build
- buildkit side changes - https://github.com/moby/buildkit/pull/580

**- How to verify it**
##### Without `--cgroup-parent` set
- First, check the initial state without this patch.`ls /sys/fs/cgroup/memory/` displays both `docker` and `buildkit` at this level.
- Rebuild the daemon with this PR and restart the docker daemon.
- Then build an arbitrary with buildkit enabled.
- `ls /sys/fs/cgroup/memory/docker` shows `buildkit` is a sub-folder.
##### With `--cgroup-parent` set
- `mkdir /sys/fs/cgroup/memory/bar`
- Restart docker daemon with `--cgroup-parent=/bar`
- Build an arbitrary with buildkit enabled.
- `ls /sys/fs/cgroup/memory/bar` shows `buildkit` is a sub-folder.

Signed-off-by: Anda Xu <anda.xu@docker.com>

### CODE REVIEW

**Comment by AntaresS** [comment]
ping @tiborvass

**Comment by AntaresS** [comment]
@AkihiroSuda @tonistiigi 
@tiborvass and I were discussing whether to propagate `cgroup-parent` from dockerd to buildkit or keep the current implementation which creates a standalone cgroup using separate namespace `buildkit`. Specifically, the two options we are thinking about:
- buildkit should use the `cgroup-parent` set from docerkd.
  - if `--cgroup-parent=/foo`, then buildkit should have a cgroup path `/sys/fs/cgroup/memory/foo/buildkit` instead of `/sys/fs/cgroup/memory/buildkit` at top level.
  - if `--cgroup-parent` is not set buildkit should use the default dockerd parent and set its cgroup to `docker/buildkit`
- Or the current implementation setting to `/sys/fs/cgroup/memory/buildkit/` is fine as is. 

Any thoughts, comments?

**Comment by tiborvass** [comment]
@AntaresS I would agree with the first proposal. If we do `foo/buildkit` then we should do `docker/buildkit` in case cgroup-parent is not specified. If we decide to just use the `foo` cgroup-parent provided, then I think we should probably use `docker` cgroup directly instead of nesting it inside `docker`.

I'm voting for nesting because it still honors the user's request and allows for even greater customization for those who really want to.

**Comment by AntaresS** [comment]
lgty too on moby side? @tonistiigi

**Comment by tonistiigi** [comment]
> if --cgroup-parent=/foo, then buildkit should have a cgroup path /sys/fs/cgroup/memory/foo/buildkit

How does the current code handle that.

> &nbsp;

We need to support the systemd style values as well.

**Comment by AntaresS** [comment]
> How does the current code handle that.

@tonistiigi Currently, by default, dockerd uses `/sys/fs/cgroup/memory/docker` and buidlkit uses `/sys/fs/cgroup/memory/buildkit`. The `--cgroup-parent` flag/option from dockerd cli only changes the daemon cgroup namespace, buildkit will still use a separate dir.

> We need to support the systemd style values as well.

Not sure which line you are quoting.

**Comment by tonistiigi** [comment]
> dockerd cli only changes the daemon cgroup namespace, buildkit will still use a separate dir.

from https://github.com/moby/buildkit/pull/580/files#diff-0094c7454519259d73efe2ea488bc27cR182 it seems to set the full value, not namespace it. In this pr I don't see prefixing as well.

> Not sure which line you are quoting.

https://github.com/moby/moby/blob/master/daemon/oci_linux.go#L696-L713

**Comment by AntaresS** [comment]
@tonistiigi I think according to https://github.com/moby/buildkit/blob/f047d8cf50979a6ac7b96d93d0ff8f852fcf6dea/executor/oci/spec_unix.go#L34-L36
the CgroupPath is namespaced in buildkit since it uses containerd. Eventually it does the same thing for non-systemd style here 
https://github.com/moby/buildkit/blob/master/vendor/github.com/containerd/containerd/oci/spec_unix.go#L79:6

And yes agreed to support systemd style.

**Comment by tonistiigi** [comment]
@AntaresS That only applies when you don't set the `CgroupPath` manually.

**Comment by AntaresS** [comment]
#### Changes since last commit 
- add support for systemd cgoup driver

File: cmd/dockerd/daemon.go [hunk 1]
```diff
@@ -285,12 +285,25 @@ func newRouterOptions(config *config.Config, daemon *daemon.Daemon) (routerOptio
 	if err != nil {
 		return opts, err
 	}
+	cgroupParent := "docker"
+	scopePrefix := "docker:buildkit"
```
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: Are you sure this works? I don't think it is allowed to have an extra `:` here

File: builder/builder-next/builder.go [hunk 2]
```diff
@@ -40,6 +40,7 @@ type Opt struct {
 	NetnsRoot         string
 	Dist              images.DistributionServices
 	NetworkController libnetwork.NetworkController
+	CgroupParent      string
```
Comment (tiborvass) [inline builder/builder-next/builder.go]: @AntaresS Let's make this DefaultCgroupParent like in buildkit.

**Comment by thaJeztah** [comment]
@AntaresS build is failing;

```
20:30:34 builder/builder-next/executor_unix.go:31:3:warning: unknown field CgroupParent in struct literal (gosimple)
20:30:34 builder/builder-next/executor_unix.go:31:3:warning: unknown field CgroupParent in struct literal (interfacer)
20:30:34 cmd/dockerd/daemon.go:289:16:warning: invalid operation: daemon (variable of type *github.com/docker/docker/daemon.Daemon) has no field or method UsingSystemd (interfacer)
20:30:34 builder/builder-next/executor_unix.go:31:3:warning: unknown field CgroupParent in struct literal (unconvert)
20:30:34 cmd/dockerd/daemon.go:289:16:warning: invalid operation: daemon (variable of type *github.com/docker/docker/daemon.Daemon) has no field or method UsingSystemd (unconvert)
```

**Comment by AntaresS** [comment]
@thaJeztah Yup. It failed in the right way. It's dependency - moby/buildkit#580 just got merged half an hour ago. I've updated the vendor. Let's see if the CI is happier now.

**Comment by AntaresS** [comment]
@seemethere It seems like the windows node doesn't have the updated code (`experimental` built it correctly). Could you help reach the node check the reason?

cc @andrewhsu

File: cmd/dockerd/daemon_unix.go [hunk 3]
```diff
@@ -107,3 +110,25 @@ func wrapListeners(proto string, ls []net.Listener) []net.Listener {
 	}
 	return ls
 }
+
+func newBuildKit(config *config.Config, d *daemon.Daemon, sm *session.Manager) (*buildkit.Builder, error) {
```
Comment (tonistiigi) [inline cmd/dockerd/daemon_unix.go]: Do not duplicate `newBuildKit` that has the same content but separate the `cgroupParent` detection to its own func.

File: cmd/dockerd/daemon_unix.go [hunk 4]
```diff
@@ -107,3 +109,18 @@ func wrapListeners(proto string, ls []net.Listener) []net.Listener {
 	}
 	return ls
 }
+
+func newCgroupParent(config *config.Config, d *daemon.Daemon, sm *session.Manager) string {
```
Comment (tonistiigi) [inline cmd/dockerd/daemon_unix.go]: why are daemon and sessionmanager passed to this function

Comment (AntaresS) [reply cmd/dockerd/daemon_unix.go]: because I forgot to remove them 😓

**Review by tiborvass — APPROVED** [review]
LGTM

**Review by tonistiigi — APPROVED** [review]

**Pull request merged**